src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -0,0 +1,382 @@
+                    {
+                        throw new RuntimeException(e);
+                    }
+                }
+
+                return (data.sizeOnHeap() - initialSizeOnHeap) + (data.sizeOffHeap() - initialSizeOffHeap) + (primaryKeysReducer.heapAllocations() - reducerHeapSize);
+            }
+            finally
+            {
+                analyzer.end();
+            }
+        }
+    }
+
+    /**
+     * Search for an expression in the in-memory index within the {@link AbstractBounds} defined
+     * by keyRange. This can either be a exact match or a range match.
+
+     * @param expression the {@link Expression} to search for
+     * @param keyRange the {@link AbstractBounds} containing the key range to restrict the search to
+     * @return a {@link RangeIterator} containing the search results
+     */
+    public RangeIterator search(Expression expression, AbstractBounds<PartitionPosition> keyRange)
+    {
+        if (logger.isTraceEnabled())
+            logger.trace("Searching memtable index on expression '{}'...", expression);
+
+        switch (expression.getOp())
+        {
+            case EQ:
+            case CONTAINS_KEY:
+            case CONTAINS_VALUE:
+                return exactMatch(expression);
+            case RANGE:
+                return rangeMatch(expression, keyRange);
+            default:
+                throw new IllegalArgumentException("Unsupported expression: " + expression);
+        }
+    }
+
+    /**
+     * Returns an {@link Iterator} over the entire dataset contained in the trie. This is used
+     * when the index is flushed to disk.
+     *
+     * @return the iterator containing the trie data
+     */
+    public Iterator<Pair<ByteComparable, PrimaryKeys>> iterator()
+    {
+        Iterator<Map.Entry<ByteComparable, PrimaryKeys>> iterator = data.entrySet().iterator();
+        return new Iterator<Pair<ByteComparable, PrimaryKeys>>()
+        {
+            @Override
+            public boolean hasNext()
+            {
+                return iterator.hasNext();
+            }
+
+            @Override
+            public Pair<ByteComparable, PrimaryKeys> next()
+            {
+                Map.Entry<ByteComparable, PrimaryKeys> entry = iterator.next();
+                return Pair.create(decode(entry.getKey()), entry.getValue());
+            }
+        };
+    }
+
+    public ByteBuffer getMinTerm()
+    {
+        return minTerm;
+    }
+
+    public ByteBuffer getMaxTerm()
+    {
+        return maxTerm;
+    }
+
+    private void setMinMaxTerm(ByteBuffer term)
+    {
+        assert term != null;
+
+        minTerm = TypeUtil.min(term, minTerm, indexContext.getValidator());
+        maxTerm = TypeUtil.max(term, maxTerm, indexContext.getValidator());
+    }
+
+    private ByteComparable encode(ByteBuffer input)
+    {
+        return isLiteral ? version -> append(ByteSource.of(input, version), ByteSource.TERMINATOR)
```

Comment:
Do we have to do this to make sure, for instance, `apple` sorts before `apples` in the trie?

Reply:
The `TERMINATOR` ensures a couple of things. It ensures that there are no prefixes in the indexes and it ensures that, with multi-component types we can search for fewer components.

I believe that this is in place to specifically exclude prefix searches. 

As it happens, if I remove the `TERMINATOR` all the in-memory tests pass but they start to fail in the on-disk tests.

I need to dig further into this but will leave it in place for now.